Firebase App Check Policy

Ankitaanku208 opened the pull request:
# Add Firebase App Check Policy and Merge Updates from Previous PRs

## Description
This PR introduces the new **Firebase App Check** policy under `gcp/service/firebase_app_check` and consolidates work from earlier efforts:

- ✅ Adds policies for Firebase App Check, structured according to the PDE cheatsheet.  
- ✅ Refines all `plan.json` inputs for consistency and validation.  
- ✅ Merges changes from [PR #77](https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/77/) and [PR #157](https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/157/) into a single, unified update.  
- ✅ Ensures alignment with existing Firebase modules and helper imports.  

## Testing
- Verified with `opa eval` against updated `plan.json` inputs.  
- Confirmed policies pass all checks under PDE standards.  

## Notes
This PR replaces and supersedes the above two PRs, providing one clean, consolidated branch for Firebase App Check.

Ankitaanku208 committed:
ALL Changes Commited

Ankitaanku208 commented:
Hi @jashan-lefty  & @jerryd10  ,I've added the branch as discussed and raised a new PR Please review and let me know if any further changes are needed.

Pat-Stuart commented on policies/gcp/firebase/auth_anonymous/disable_anonymous/policy.rego:
```diff
@@ -0,0 +1,22 @@
+    {
+        "condition": "Check if allow_anonymous is false",
+        "attribute_path" : ["values","triggers","allow_anonymous"],
+        "values" : ["false"],
```
This should likely just be a bool not a string

Pat-Stuart commented on policies/gcp/firebase/auth_anonymous/vars.rego:
```diff
@@ -0,0 +1,7 @@
+variables := {
+    "friendly_resource_name": "Firebase Authentication",
+    "resource_type": "google_identity_platform_project_default_config",
+    "resource_value_name": ["values","triggers", "name"]
```
You can't have multiple resource_value_name

Pat-Stuart commented on inputs/gcp/firebase/functions_region/nc.tf:
```diff
@@ -0,0 +1,12 @@
+# Non-compliant Firebase Function deployed in disallowed region
+# Keep "nc" as the name to indicate that this resource and its attributes are non-compliant
+
+resource "null_resource" "nc" {
```
Why are we using a null resource?

Pat-Stuart commented on inputs/gcp/firebase/firestore_rules/c.tf:
```diff
@@ -0,0 +1,20 @@
+# Compliant Firestore ruleset requiring authentication
+# Keep "c" as the name to indicate that this resource and its attributes are compliant
+
+resource "google_firestore_ruleset" "c" {
```
This resource doesn't seem to exist and definitely isn't a part of your service

Pat-Stuart commented on inputs/gcp/firebase/storage/c.tf:
This has nothing to do with your service

Pat-Stuart commented on policies/gcp/firebase/firebase_project_rate_limit/rate_limit/policy.rego:
```diff
@@ -0,0 +1,25 @@
+        {
+            "condition": "Rate limiting must be enabled",
+            "attribute_path": ["values", "content"],
+            "values": ["rate_limit=true"],
```
This is very likely incorrect

Pat-Stuart commented on policies/gcp/firebase/firestore_rules/restricted_access/policy.rego:
```diff
@@ -0,0 +1,22 @@
+    {
+        "condition": "Check if rules contain 'allow read, write: if true'",
+        "attribute_path" : ["values", "source", "files", 0, "content"],
+        "values" : ["allow read, write: if true"],
```
This is very likely incorrect

Pat-Stuart reviewed (changes requested):
There is lots of big issues here. 

You have used null_resources for the majority of your Terraform which we do not want and effectively doesn't test your service resources. 

Many attribute paths and values are likely incorrect however you would be matching against null resources so you don't know what it is actually going to look like on actual resources.

You have a few resources and policies that are unrelated to your service.

Many resource_value_names are incorrect as we can't use an array for what should be a string identifying an attribute.

Commit:
fixed auth_anonymous and auth_provider with opa tested (screenshot atttached)

Ankitaanku208 commented:
Hi @Pat-Stuart Could you please check again i have fixed all null resouces to valid resources and tested out polices.
<img width="1012" height="216" alt="image" src="https://github.com/user-attachments/assets/5c3ebc14-7d61-49fc-b10d-eb4f3e64058f" />

Pat-Stuart commented:
The image you've posted shows your policy failing to catch a non-compliant resource.
Their is still  TF unrelated to resources for your service ie. google_identity_platform_config and google_cloudfunctions_function.

Other issues with policies still exist

Commit:
Added 3 Policies with attested mysel with opa eval and removed previous policies

Ankitaanku208 commented:
Hi @Pat-Stuart,

I’ve reviewed the three new policies for the specific Firebase App Check and removed the previous ones due to multiple issues. I’m confident in these three policies. Could you please review them and let me know if there are any concerns? Once confirmed, I’ll proceed with adding more.
I have also attached screenshots for reference 
![WhatsApp Image 2025-09-24 at 2 14 23 AM](https://github.com/user-attachments/assets/13d88e2c-d814-4510-8552-da3ba3c35f19)
![WhatsApp Image 2025-09-24 at 2 10 43 AM](https://github.com/user-attachments/assets/0cf95a77-155c-4c41-8586-f8b774b3d410)
![WhatsApp Image 2025-09-24 at 2 10 41 AM](https://github.com/user-attachments/assets/cca35dec-5b43-4408-9130-63287c67a599)

Thanks!

Pat-Stuart commented on policies/gcp/firebase_app_check/app_check/app_attest/policy.rego:
We don't want naming policies so this can be removed but it works and is the right idea

Pat-Stuart commented on policies/gcp/firebase_app_check/app_check/token_ttl/policy.rego:
Good concept and the sort of policy we are looking for but it won't work if its a string ie. ends with "s", this is a limitation of the current helpers.  Consider just setting a short whitelist value "1800s" and making a note in your documentation.

Pat-Stuart replied:
My above comment still applies, the policy will pass any string value.

Pat-Stuart commented on policies/gcp/firebase_app_check/app_check/project/policy.rego:
The policy works so good job but it is basically just a naming policy so its not really what we are after.

Pat-Stuart commented on policies/gcp/firebase_app_check/app_check/vars.rego:
```diff
@@ -0,0 +1,8 @@
+variables := {
+    "friendly_resource_name": "Firebase App Check App Attest Config",
+    "resource_type": "google_firebase_app_check_app_attest_config",
+    "resource_value_name": "name",
```
I can't see it in the code given but in your screenshots I see a example like "nc_example_whatever".

Please make sure you follow the naming convention so just "c", "nc" and possible digits if needed "nc1234" is fine but "nc-bad" is not.

Pat-Stuart reviewed (commented):
The policies work which is good, just consider the security relevance when you are doing other policies.

Commit:
added security relevance firebase app policies

Ankitaanku208 commented:
I’ve added four more policies and tested them with OPA. Screenshots are attached—please review and share any feedback or suggested changes. 
![WhatsApp Image 2025-09-24 at 10 45 44 AM](https://github.com/user-attachments/assets/d972e22a-54d5-45fe-96c1-684f5b8ce990)
![WhatsApp Image 2025-09-24 at 10 46 17 AM](https://github.com/user-attachments/assets/1d90d936-a318-44b7-aea9-d4bd1bcd62b1)
![WhatsApp Image 2025-09-24 at 10 47 30 AM](https://github.com/user-attachments/assets/ff5bcc3e-9d22-48c2-ae3e-af7baac53833)
![WhatsApp Image 2025-09-24 at 10 49 41 AM](https://github.com/user-attachments/assets/1cb04f58-85c3-4c10-ba83-8c443a1388fb)
thanks

Commit:
Tested with opa and checked with security relevance 3 policies more addded

Ankitaanku208 commented:
I've included three additional policies focusing on security relevance and verified them for any issues. Screenshots are attached for reference. Everything should be ready from my side for the next steps.
![WhatsApp Image 2025-09-24 at 12 18 01 PM](https://github.com/user-attachments/assets/9c1e7efd-4370-4d29-b8a0-804265cee2a1)
![WhatsApp Image 2025-09-24 at 12 22 45 PM](https://github.com/user-attachments/assets/4f3ff157-706a-4c6a-ab61-855d401f7f67)
@
![WhatsApp Image 2025-09-24 at 12 30 01 PM](https://github.com/user-attachments/assets/d7c6c687-539f-4ed7-af5f-b4b22709fdbd)
thanks.

Pat-Stuart commented on policies/gcp/firebase_app_check/android_app/sha256_hashes_presence/policy.rego:
This policy wont work as a limitation of the current helpers, you can see that it fails to catch any resources in your current screenshot. I would make a note in documentation and remove

Pat-Stuart commented on policies/gcp/firebase_app_check/site/site_id_pattern/policy.rego:
```diff
@@ -0,0 +1,24 @@
+        {
+            "condition": "Check if site_id matches the pattern",
+            "attribute_path": ["site_id"],
+            "values": ["*", [["^[a-z0-9-]+$"]]],
```
We don't support straight Regex, the policy type is more of a find and replace. You can see that this fails both resources in your screenshots

Pat-Stuart commented on policies/gcp/firebase_app_check/debug_token/display_name/policy.rego:
This is a naming policy and can be removed

Pat-Stuart commented on policies/gcp/firebase_app_check/recaptcha_v3_config/token_ttl_presence/policy.rego:
I would probably just whitelist a reasonable value ie. "1800s", we are allowing anything that is input here.

Pat-Stuart commented on policies/gcp/firebase_app_check/play_integrity_config/token_ttl_presence/policy.rego:
I would probably just whitelist a reasonable value ie. "1800s", we are allowing anything that is input here.

Pat-Stuart reviewed (changes requested):
Sorry I left this pending this morning, close but again a lot of policies are naming or don't enforce a secure standard

Commit:
removed 256 sha and documented it

Commit:
as per suggestion of 1800s

Ankitaanku208 commented:
> Sorry I left this pending this morning, close but again a lot of policies are naming or don't enforce a secure standard
Hi @Pat-Stuart 
Could you please elaborate about the security relevenace so i can fix the policies according to it and working on site_id_pattern and token_ttl for app check and other than that all policies are okay or not let me know so i wont work on that

Ankitaanku208 committed:
fixed error

Commit:
sha256 firebase policy

Commit:
deleted 1800s policies for issues not resolvinng in them

Commit:
deleted anbd checked three policies c which is working fine

Ankitaanku208 commented:
I’ve fixed the issues. Can you test the changes and let me know? I deleted several policies I couldn’t resolve. and yes i've tested with opa results too plus they are security relevant policies

Commit:
tested api_key policy and debug_token with opa and correct results

Commit:
fixed sha256 null issue

Commit:
fixed the recaptcha policy

Commit:
firebase app check documentation
